src/concepts/mods.md

```diff
@@ -1 +1,506 @@
+_modularity_. As alluded to previously in this chapter, we have mentioned how we
+would like to be able to _swap out_ certain implementations for others as
+necessary, to enforce clean boundaries of code dependency and produce more
+maintainable code. As it stands, we have only seen modules used for conveniently
+wrapping our code, however in this section we are going to discuss how modules
+allow us to maintain multiple implementations of the same specification.
+
+Consider the following signature for performing geometric operations on the
+Cartesian plane:
+
+```sml
+signature GEOMETRY =
+sig
+    type point
+
+    val origin : point
+    val rotate : point -> real -> point
+
+    val get_x : point -> real
+    val get_y : point -> real
+    val get_dist : point -> real
+end
+```
+
+Suppose that we are interested in performing operations on points within the
+Cartesian plane. Then, we may be interested in a structure ascribing to this
+signature. The question then becomes - how should this signature be implemented?
+We have previously discussed swapping out more code for more efficient
+implementations, however in that case we clearly consider one unilaterally
+superior to the other.
+
+In this signature, note that we have left the type of `point` _abstract_, meaning that it
+is up to the structure to define what precisely the type of `point` is - it is
+not decided by the signature. In this case, there are at least two discrete ways
+that we can consider points on the Cartesian plane - as rectangular or polar coordinates.
+
+Consider the following structures ascribing to the `GEOMETRY` signature:
+
+```sml
+structure RectGeo : GEOMETRY =
+struct
+    (* pair of (x, y) coordinates *)
+    type point = real * real
+
+    val origin = (0.0, 0.0)
+    fun rotate (x, y) r = (x * Math.cos r) + (y * Math.sin r)
+
+    fun get_x (x, y) = x
+    fun get_x (x, y) = y
+    fun get_dist (x, y) = Math.sqrt (Math.pow (x, 2.0) + Math.pow (y, 2.0))
+end
+
+```sml
+structure PolarGeo : GEOMETRY =
+struct
+    (* pair of (d, theta), where d is distance from origin and theta is angle from 0 degrees *)
+    type point = real * real
+
+    (* theoretically here theta could be anything *)
+    val origin  = (0.0, 0.0)
+    fun rotate (d, theta) r = (d, theta + r)
+
+    fun get_x (d, theta) = d * (Math.cos theta)
+    fun get_y (d, theta) = d * (Math.sin theta)
+    fun get_dist (d, theta) = d
+end
+```
+
+It should be relatively clear that these are two perfectly valid ways of
+representing an (albeit limited) implementation of coordinate geometry. Both
+have their own strengths and weaknesses, such as rectangular coordinates having
+more convenient access to computing the x and y coordinates of a given point,
+and polar coordinates more easily rotating points about the origin and
+calculating the distance of a point from the origin. As such, it comes down to
+_context_ to determine which should be used in any given circumstance.
```

Comment:
I am personally not 100% convinced by this example. In both structures, the client can get x and y coordinates, find the distance, and use the rotate function (with the same ease). I guess the only difference is that, since it's transparently ascribed, the client can peek at the representation of the point and do things with it that aren't specified in the structure. That feels kind of janky though (and wasn't mentioned).

I have an idea (although idk if it's much better): have something that implements the signature simply, and another more verbose version that also prints out information when the values and functions are used. although that uses side effects which might be messy